artiq/firmware/runtime/rtio_dma.rs

```diff
@@ -1,18 +1,183 @@
+            let _lock = ddma_mutex.lock(io).unwrap();
+            if clock::get_ms() < max_time {
+                let traces = unsafe { TRACES.get(&id).unwrap() };
+                for (_dest, trace) in traces {
+                    match trace.state {
+                        RemoteState::PlaybackEnded {error: _, channel: _, timestamp: _} => (),
+                        _ => return false
+                    }
+                }
+            }
+            true
+        }).unwrap();
+        if clock::get_ms() > max_time {
+            error!("Remote DMA await done timed out");
+            return Err("Timed out waiting for results.");
+        }
+        // clear the internal state, and if there have been any errors, return one of them
+        let mut playback_state: RemoteState = RemoteState::PlaybackEnded { error: 0, channel: 0, timestamp: 0 };
+        {
+            let _lock = ddma_mutex.lock(io).unwrap();
+            let traces = unsafe { TRACES.get_mut(&id).unwrap() };
+            for (_dest, trace) in traces {
+                match trace.state {
+                    RemoteState::PlaybackEnded {error: e, channel: _c, timestamp: _ts} => if e != 0 { playback_state = trace.state.clone(); },
+                    _ => (),
+                }
+                trace.state = RemoteState::Loaded;
+            }
+        }
+        Ok(playback_state)
+    }
+
+    pub fn erase(io: &Io, aux_mutex: &Mutex, routing_table: &RoutingTable,
+            ddma_mutex: &Mutex, id: u32) {
+        let _lock = ddma_mutex.lock(io).unwrap();
+        let destinations = unsafe { TRACES.get(&id).unwrap() };
+        for destination in destinations.keys() {
+            match drtio::ddma_send_erase(io, aux_mutex, routing_table, id, *destination) {
+                Ok(_) => (),
+                Err(e) => error!("Error erasing trace on DMA: {}", e)
+            } 
+        }
+        unsafe { TRACES.remove(&id); }
+    }
+
+    pub fn upload_traces(io: &Io, aux_mutex: &Mutex, routing_table: &RoutingTable,
+            ddma_mutex: &Mutex, id: u32) {
+        let _lock = ddma_mutex.lock(io);
+        let traces = unsafe { TRACES.get_mut(&id).unwrap() };
+        for (destination, mut trace) in traces {
+            match drtio::ddma_upload_trace(io, aux_mutex, routing_table, id, *destination, trace.get_trace())
+            {
+                Ok(_) => trace.state = RemoteState::Loaded,
+                Err(e) => error!("Error adding DMA trace on destination {}: {}", destination, e)
+            }
+        }
+    }
+
+    pub fn playback(io: &Io, aux_mutex: &Mutex, routing_table: &RoutingTable,
+            ddma_mutex: &Mutex, id: u32, timestamp: u64) {
+        // triggers playback on satellites
+        let destinations = unsafe { 
+            let _lock = ddma_mutex.lock(io).unwrap();
+            TRACES.get(&id).unwrap() };
+        for (destination, trace) in destinations {
+            {
+                let _lock = ddma_mutex.lock(io).unwrap();
```

Comment:
ddma lock must be dropped for the drtio send playback function - in case we receive a PlaybackStatus in the meanwhile and need to update the state. To prevent a deadlock, a lock is created briefly to verify the internal state on every loop iteration, and then dropped.